src/Infrastructure/Search/FilterableTermsUpdater.php

```diff
@@ -222,48 +222,177 @@ private function maybeCreateTerms(
+        MultilingualString|MultilingualLink|AdmissionRequirement|Degree $flatProperty,
+        array &$taxonomiesCache,
+        string $taxonomy
+    ): ?TermData {
+
+        $termIdsData = $this->termIdsData($flatProperty, $taxonomiesCache[$taxonomy]);
+
+        if (is_null($termIdsData)) {
+            return null;
+        }
+
+        [$termId, $remoteTermId] = $termIdsData;
+
+        return TermData::fromArray([
+            'taxonomy' => $taxonomy,
+            'name' => $flatProperty instanceof MultilingualString
+                ? $flatProperty
+                : $flatProperty->name(),
+            'slug' => $this->retrieveTermSlug($flatProperty),
+            'term_id' => $termId,
+            'remote_term_id' => $remoteTermId,
+            'parent_term_id' => $this->retrieveParentTermId(
+                $flatProperty,
+                $taxonomiesCache[$taxonomy]
+            ),
+        ]);
+    }
+
+    /**
+     * @param AdmissionRequirement|Degree|MultilingualLink|MultilingualString $flatProperty
+     * @param array $taxonomyCache
+     * @psalm-param array<int, array{name?: string, original_id: int}> $taxonomyCache
+     * @return array<int>|null
+     */
+    private function termIdsData(
+        MultilingualString|MultilingualLink|AdmissionRequirement|Degree $flatProperty,
+        array &$taxonomyCache
+    ): ?array {
+
+        $termId = $this->idGenerator->termIdsList($flatProperty)[0] ?? null;
+
+        if (is_null($termId)) {
+            return null;
+        }
+
+        $name = $flatProperty instanceof MultilingualString ? $flatProperty : $flatProperty->name();
+        $newTermId = self::arraySearchBy(
+            $taxonomyCache,
+            self::TAXONOMIES_CACHE_ORIGINAL_ID_PROPERTY,
+            $termId
+        );
+
+        if (is_int($newTermId)) {
+            // Term was persisted already
+            return [$newTermId, $termId];
+        }
+
+        $newTermId = self::arraySearchBy(
+            $taxonomyCache,
+            self::TAXONOMIES_CACHE_NAME_PROPERTY,
+            $name->inGerman()
+        );
+
+        if (is_int($newTermId)) {
+            // Term with the name already exists
+            $taxonomyCache[$newTermId] = [
+                self::TAXONOMIES_CACHE_ORIGINAL_ID_PROPERTY => $termId,
+            ];
+
+            return [$newTermId, $termId];
+        }
+
+        return [0, $termId];
+    }
+
+    /**
+     * @param AdmissionRequirement|Degree|MultilingualLink|MultilingualString $flatProperty
+     * @param array $taxonomyCache
+     * @psalm-param array<int, array{name?: string, original_id: int}> $taxonomyCache
+     * @return int
+     */
+    private function retrieveParentTermId(
+        MultilingualString|MultilingualLink|AdmissionRequirement|Degree $flatProperty,
+        array &$taxonomyCache
+    ): int {
+
+        if (!$flatProperty instanceof Degree && !$flatProperty instanceof AdmissionRequirement) {
+            return 0;
+        }
```

Comment:
Does extracting the condition in something like `isHierarchicalPropery` make sense?